cpp/include/cudf/dictionary/detail/iterator.cuh

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column_device_view.cuh>
+
+namespace cudf {
+namespace dictionary {
+namespace detail {
+
+template <typename KeyType>
+struct dictionary_access_fn {
+  dictionary_access_fn(column_device_view const& d_dictionary) : d_dictionary{d_dictionary} {}
+
+  __device__ KeyType operator()(size_type idx) const
+  {
+    if (d_dictionary.is_null(idx)) return KeyType{};
+    auto keys = d_dictionary.child(dictionary_column_view::keys_column_index);
+    return keys.element<KeyType>(static_cast<size_type>(d_dictionary.element<dictionary32>(idx)));
+  };
+
+ private:
+  column_device_view const d_dictionary;
+};
+
+/**
+ * @brief Create dictionary iterator that produces key elements.
+ *
+ *
+ * @throw cudf::logic_error if `dictionary_column` is not a dictionary column.
+ *
+ * @param dictionary_column The dictionary device view to iterate.
+ * @return Iterator
+ */
+template <typename KeyType>
+auto make_dictionary_iterator(column_device_view const& dictionary_column)
```

Comment:
Should this variant should only support columns w/o nulls? I'd imagine we'd have a `pair_iterator` equivalent that can be used for columns with null elements.

Reply:
I was planning to add the pair-iterator here once I need it. Should be soon but not in this PR.

Reply:
Makes sense. I'd think for this PR that this iterator should `CUDF_EXPECT` that the input column doesn't contain nulls. I think similar iterators do the same thing.

Reply:
Yes, good point. It also seems a little weak on documentation too now that I look a back at it.

Reply:
The `column_device_view` does not have a `null_count()` nor a `has_nulls()` member so this utility is unable to check. The iterator handles nulls either way so the caller can do the check if necessary.